### PR TITLE
Fix custom ticks

### LINKED ISSUE #812: Bar chart label overlap
#### I am submitting a...

- [ ] Feature request
- [X] Design defect
- [ ] Source code defect
- [ ] Demo/documentation defect
- [ ] Other

#### charts version:
```json
"@carbon/charts": "0.38.2",
"@carbon/charts-react": "0.38.2",
```

### Issue description
When the time scale dataset is reduced from multiple days to a single day (using custom ticks) the x-axis displays an overlapped text

#### Steps to produce the issue
1. Use custom ticks;
2. Set the dataset to at least 7 days (ticks);
3. Change the chart state using a dataset that contains a single day (tick);

#### Current behavior
The text (date and time) on the x-axis is overlapped;

#### Expected behavior
It should be displayed only the label for the current day;

### Screenshot or recording
7 Days:
image

Single day:
image

### PR DESCRIPTION
**Pull request by natashadecoste**
Closes #812

### CODE REVIEW

File: packages/core/src/components/axes/axis.ts [hunk 1]
```diff
@@ -295,8 +295,25 @@ export class Axis extends Component {
 
 		// prioritize using a custom array of values rather than number of ticks
 		// if both are provided. custom tick values need to be within the domain
+		const bounds = this.services.cartesianScales.getScaleByPosition(axisPosition).domain();
```
Comment (theiliad) [inline packages/core/src/components/axes/axis.ts:298]:
```suggestion
		const [lowerBound, upperBound] = this.services.cartesianScales.getScaleByPosition(axisPosition).domain();
```

File: packages/core/src/components/axes/axis.ts [hunk 2]
```diff
@@ -295,8 +295,25 @@ export class Axis extends Component {
...
+
 		if (userProvidedTickValues) {
-			axis.tickValues(userProvidedTickValues);
+			if (isTimeScaleType) {
+				// check the supplied ticks are within the domain
+				const validTicks = userProvidedTickValues.filter((tick) => {
+					const result = new Date(tick).getTime();
+					// if its within the range use it
+					if (result >= new Date(lowerBound).getTime() && result <= new Date(upperBound).getTime()) {
+						return tick;
```
Comment (theiliad) [inline packages/core/src/components/axes/axis.ts:309]:
```suggestion
						return result >= new Date(lowerBound).getTime() && result <= new Date(upperBound).getTime();
```

Comment (theiliad) [reply packages/core/src/components/axes/axis.ts:309]: filter doesn't need the actual item returned, just a boolean. We should get rid of the if-statement

File: packages/core/src/components/axes/axis.ts [hunk 3]
```diff
@@ -295,8 +295,25 @@ export class Axis extends Component {
...
+		const bounds = this.services.cartesianScales.getScaleByPosition(axisPosition).domain();
+		const lowerBound = bounds[0];
+		const upperBound = bounds[1];
+
 		if (userProvidedTickValues) {
-			axis.tickValues(userProvidedTickValues);
+			if (isTimeScaleType) {
+				// check the supplied ticks are within the domain
+				const validTicks = userProvidedTickValues.filter((tick) => {
+					const result = new Date(tick).getTime();
```
Comment (theiliad) [inline packages/core/src/components/axes/axis.ts:306]:
```suggestion
					const tickTimestamp = new Date(tick).getTime();
```

File: packages/core/src/components/axes/axis.ts [hunk 4]
```diff
@@ -295,8 +295,25 @@ export class Axis extends Component {
...
+					// if its within the range use it
+					if (result >= new Date(lowerBound).getTime() && result <= new Date(upperBound).getTime()) {
+						return tick;
+					}
+				});
+				axis.tickValues(validTicks);
+			} else {
+				const validTicks = userProvidedTickValues.filter((tick) => tick >= lowerBound && tick <= upperBound);
```
Comment (theiliad) [inline packages/core/src/components/axes/axis.ts:314]: what if this is a `discrete` scale?

**Review by theiliad — CHANGES_REQUESTED** [review]

File: packages/core/src/components/axes/axis.ts [hunk 5]
```diff
@@ -294,9 +294,24 @@ export class Axis extends Component {
...
-			axis.tickValues(userProvidedTickValues);
+			if (isTimeScaleType) {
+				// check the supplied ticks are within the time domain
+				validTicks = userProvidedTickValues.filter((tick) => {
+					const tickTimestamp = new Date(tick).getTime();
+					return tickTimestamp >= new Date(lowerBound).getTime() && tickTimestamp <= new Date(upperBound).getTime();
+				});
+			} else if (axisScaleType === ScaleTypes.LABELS) {
+				const discreteDomain = this.services.cartesianScales.getScaleByPosition(axisPosition).domain();
+				validTicks = userProvidedTickValues.filter((tick) => discreteDomain.includes(tick));
```
Comment (theiliad) [inline packages/core/src/components/axes/axis.ts:309]: 💯

File: packages/core/demo/data/line.ts [hunk 6]
```diff
@@ -228,7 +228,10 @@ export const lineTimeSeriesCustomDomainOptions = {
 			title: "2019 Annual Sales Figures",
 			domain: [new Date(2019, 0, 5), new Date(2019, 0, 15)],
 			mapsTo: "date",
-			scaleType: "time"
+			scaleType: "time",
+			ticks: {
+				values: [new Date(2019, 0, 11)]
+			}
```
Comment (theiliad) [inline packages/core/demo/data/line.ts:234]: do we want to keep this?

**Review by theiliad — DISMISSED** [review]

File: packages/core/demo/data/line.ts (hunk 6 above)
Comment (natashadecoste) [reply packages/core/demo/data/line.ts:234]: yeah we dont need to keep this - ill remove this

**Review by theiliad — APPROVED** [review]

**Review by sophiiae — APPROVED** [review]

**Pull request merged**